src/gluonts/model/tft/_engine.py

```diff
@@ -0,0 +1,315 @@
+                net=net,
+                hybridize=self.hybridize,
+                static_alloc=True,
+                static_shape=True,
+            ):
+                batch_size = train_iter.batch_size
+
+                best_epoch_info = {
+                    "params_path": f"{base_path()}-init.params",
+                    "epoch_no": -1,
+                    "score": np.Inf,
+                }
+
+                lr_scheduler = MetricAttentiveScheduler(
+                    objective="min",
+                    patience=self.patience,
+                    decay_factor=self.learning_rate_decay_factor,
+                    min_lr=self.minimum_learning_rate,
+                )
+
+                optimizer = mx.optimizer.Adam(
+                    learning_rate=self.learning_rate,
+                    lr_scheduler=lr_scheduler,
+                    wd=self.weight_decay,
+                    clip_gradient=self.clip_gradient,
+                )
+
+                trainer = MXTrainer(
+                    net.collect_params(),
+                    optimizer=optimizer,
+                    kvstore="device",  # FIXME: initialize properly
+                )
+
+                def loop(
+                    epoch_no: int,
+                    batch_iter: DataLoader,
+                    is_training: bool = True,
+                ) -> mx.metric.Loss:
+                    tic = time.time()
+                    epoch_loss = mx.metric.Loss()
+                    with tqdm(batch_iter) as it:
+                        for batch_no, data_entry in enumerate(it, start=1):
+                            if self.halt:
+                                break
+                            args = inspect.signature(
+                                net.hybrid_forward
+                            ).parameters
+                            inputs = []
+                            for n, (name, arg) in enumerate(args.items()):
+                                if n == 0:
+                                    if name == "F":
+                                        continue
+                                    else:
+                                        raise RuntimeError(
+                                            f"Expected first argument of HybridBlock to be `F`, "
+                                            f"but found `{name}`"
+                                        )
+                                if name in data_entry:
+                                    inputs.append(data_entry[name])
+                                elif not (arg.default is inspect._empty):
+                                    inputs.append(arg.default)
+                                else:
+                                    raise RuntimeError(
+                                        f"The value of argument `{name}` of HybridBlock is not provided, "
+                                        f"and no default value is set."
+                                    )
```

Comment:
Some observations:
* The code here seems like a refined version of [this](https://github.com/awslabs/gluon-ts/blob/master/src/gluonts/mx/trainer/_base.py#L250)-plus-[this](https://github.com/awslabs/gluon-ts/blob/df1e50d462f97ac2bf8ee5865f165cce2c066d83/src/gluonts/support/util.py#L169-L176) allowing for defaults;
* This is duplicated in the forecast generator defined below;
* Both in the trainer and forecast generator now have an obsolete `input_names` constructor argument.

If you really want to push for this mechanism for TFT, my suggestion here is to propose this new mechanism in a separate PR where you showcase it with a minimal example model. This way you can focus on getting this into the default `Trainer` class and `ForecastGenerator` types **in a backward compatible way** (so that other models keep working fine with the default `Trainer` and `ForecastGenerator` classes).

Then you should then be able to proceed with the current PR, but avoiding a lot of code duplication.

Reply:
This of course applies to #961 as well, so the code duplication savings double :-)